pkg/querier/blocks_store_queryable.go

```diff
@@ -658,12 +658,16 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(
 						}
 					}
 					chunksSize := countChunkBytes(s)
+					dataSize := countDataBytes(s)
 					if chunkBytesLimitErr := queryLimiter.AddChunkBytes(chunksSize); chunkBytesLimitErr != nil {
 						return validation.LimitError(chunkBytesLimitErr.Error())
 					}
 					if chunkLimitErr := queryLimiter.AddChunks(len(s.Chunks)); chunkLimitErr != nil {
 						return validation.LimitError(chunkLimitErr.Error())
 					}
+					if dataBytesLimitErr := queryLimiter.AddDataBytes(dataSize); dataBytesLimitErr != nil {
```

Comment:
One question here. So the query limiter we are using have the same lifecycle with each query, let's say if we first query a store gateway and we limit some data bytes, then maybe due to network issue or SG restart, the stream erros and we need to retry another store gateway.
In this case, do you think it makes more sense to release the bytes we consumed and then start retrying another store gateway?

Reply:
This is a valid point. I also want to note that we are not doing this for other existing limits like fetched chunks and fetched series limits.

The query-frontend only retries 5XXs. In this case, we'll be returning a 422 error which will not be retried.

Reply:
We could do something like:

```
if isRetryableError(err) {
  level.Warn(spanLog).Log("err", errors.Wrapf(err, "failed to receive series from %s due to retryable error", c.RemoteAddress()))
  queryLimiter.RemoveSeries(seriesCnt)
  queryLimiter.RemoveDataBytes(databytesCnt)
  queryLimiter.RemoveChunkBytes(chunkbytesCnt)
  return nil
}
```

If others agree with this approach, I can implement this as an enhancement in another PR.

Reply:
SGTM

Reply:
That's a good point. We should follow up with another PR to fix this i think!